### PR TITLE
Update Committee Helper Part 2

### PR DESCRIPTION
**Pull request by terencechain**
This PR is part of #2307 and a follow up to #2354, it aims to align committee related functions up to spec. It's going to be a messy PR but necessary to get 0.6 off running.

Changes:
-  Removed `CrosslinkCommitteeAtSlot`
-  Removed `CurrentEpochCommitteeCount`
-  Removed `PrevEpochCommitteeCount`
-  Removed `NextEpochCommitteeCount`
-  Updated `ComputeCommittee`
-  Updated `AttestationParticipants`
-  Updated `BeaconProposerIndex`
-  Updated `AttestationData` protobuf to include `target_epoch`
-  Removed old core functions which still used `CrosslinkCommitteeAtSlot`
-  Removed old epoch processing functions which still used `CrosslinkCommitteeAtSlot`

### CODE REVIEW

File: beacon-chain/attestation/service.go [hunk 1]
```diff
@@ -234,35 +232,16 @@ func (a *Service) updateAttestation(ctx context.Context, headRoot [32]byte, beac
 		}
 	}
 
-	cachedCommittees, err = committeeCache.CommitteesInfoBySlot(slot)
+	committee, err := helpers.CrosslinkCommitteeAtEpoch(beaconState, helpers.CurrentEpoch(beaconState), attestation.Data.Shard)
 	if err != nil {
 		return err
 	}
-	if cachedCommittees == nil {
```
Comment (terencechain) [inline beacon-chain/attestation/service.go]: The committee cache is too tightly coupled with the old spec. Here we temporary remove the committee cache, will add it back once 0.6 is done

File: beacon-chain/core/balances/rewards_penalties.go [hunk 2]
```diff
@@ -310,70 +309,3 @@ func AttestationInclusion(
...
-// 		let crosslink_committees_at_slot = get_crosslink_committees_at_slot(slot).
-// 		For every (crosslink_committee, shard) in crosslink_committee_at_slot,
-// 		and every index in crosslink_committee:
-//			If index in attesting_validators(crosslink_committee),
-//			state.validator_balances[index] += base_reward(state, index) *
-//			total_attesting_balance(crosslink_committee) //
-//			get_total_balance(state, crosslink_committee)).
-//			If index not in attesting_validators(crosslink_committee),
-//			state.validator_balances[index] -= base_reward(state, index).
-func Crosslinks(
```
Comment (terencechain) [inline beacon-chain/core/balances/rewards_penalties.go]: Old, getting replaced by `2460`

File: beacon-chain/core/epoch/epoch_processing.go [hunk 3]
```diff
@@ -186,92 +186,6 @@ func ProcessRegistryUpdates(state *pb.BeaconState) *pb.BeaconState {
...
-// the state epoch and wining root.
-//
-// Spec pseudocode definition:
-//	For every slot in range(get_epoch_start_slot(previous_epoch), get_epoch_start_slot(next_epoch)),
-// 	let `crosslink_committees_at_slot = get_crosslink_committees_at_slot(state, slot)`.
-// 		For every `(crosslink_committee, shard)` in `crosslink_committees_at_slot`, compute:
-// 			Set state.latest_crosslinks[shard] = Crosslink(
-// 			epoch=slot_to_epoch(slot), crosslink_data_root=winning_root(crosslink_committee))
-// 			if 3 * total_attesting_balance(crosslink_committee) >= 2 * total_balance(crosslink_committee)
-func ProcessCrosslinks(
```
Comment (terencechain) [inline beacon-chain/core/epoch/epoch_processing.go]: old, getting replaced by #2517

File: beacon-chain/core/helpers/committee.go [hunk 4]
```diff
@@ -154,97 +119,6 @@ func ComputeCommittee(
...
-//    seed = generate_seed(state, epoch)
-//
-//    return [
-//      (
-//        compute_committee(indices, seed, committees_per_slot * offset + i, committees_per_epoch),
-//        (slot_start_shard + i) % SHARD_COUNT,
-//      )
-//      for i in range(committees_per_slot)
-//    ]
-func CrosslinkCommitteesAtSlot(state *pb.BeaconState, slot uint64) ([]*CrosslinkCommittee, error) {
```
Comment (terencechain) [inline beacon-chain/core/helpers/committee.go]: Old

File: beacon-chain/core/helpers/committee.go [hunk 5]
```diff
@@ -511,11 +329,11 @@ func CommitteeAssignment(
 // Spec pseudocode definition:
 // 	def get_shard_delta(state: BeaconState, epoch: Epoch) -> int:
 //    return min(get_epoch_committee_count(state, epoch), SHARD_COUNT - SHARD_COUNT // SLOTS_PER_EPOCH)
-func ShardDelta(beaconState *pb.BeaconState, epoch uint64) uint64 {
```
Comment (nisdas) [inline beacon-chain/core/helpers/committee.go]: maybe have another name other than state ? state is the name of one of our core packages too.

File: beacon-chain/attestation/service_test.go [hunk 6]
```diff
@@ -376,6 +380,7 @@ func TestUpdateLatestAttestation_CacheEnabledAndHit(t *testing.T) {
 }
 
 func TestUpdateLatestAttestation_InvalidIndex(t *testing.T) {
+	t.Skip()
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/attestation/service_test.go]: Fixed. No longer skipped

File: beacon-chain/attestation/service_test.go [hunk 7]
```diff
@@ -420,6 +425,7 @@ func TestUpdateLatestAttestation_InvalidIndex(t *testing.T) {
 }
 
 func TestUpdateLatestAttestation_BatchUpdate(t *testing.T) {
+	t.Skip()
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/attestation/service_test.go]: Fixed. No longer skipped

File: beacon-chain/core/epoch/epoch_processing.go [hunk 8]
```diff
@@ -186,92 +186,6 @@ func ProcessRegistryUpdates(state *pb.BeaconState) *pb.BeaconState {
...
-// Spec pseudocode definition:
-//	def process_ejections(state: BeaconState) -> None:
-//    """
-//    Iterate through the validator registry
-//    and eject active validators with balance below ``EJECTION_BALANCE``.
-//    """
-//    for index in get_active_validator_indices(state.validator_registry, current_epoch(state)):
-//        if state.validator_balances[index] < EJECTION_BALANCE:
-//            exit_validator(state, index)
-func ProcessEjections(state *pb.BeaconState, enableLogging bool) (*pb.BeaconState, error) {
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/core/epoch/epoch_processing.go]: It's no longer in the spec. It's refactored into `process_registry_updates`

File: beacon-chain/core/helpers/committee.go (hunk 5 above)
Comment (terencechain) [reply beacon-chain/core/helpers/committee.go]: renamed it back to `beaconState`

**Pull request merged**